bftengine/src/preprocessor/PreProcessor.cpp

```diff
@@ -589,24 +593,58 @@ template <typename T>
 void PreProcessor::messageHandler(MessageBase *msg) {
   T *trueTypeObj = new T(msg);
   delete msg;
-  if (validateMessage(trueTypeObj)) {
-    onMessage<T>(trueTypeObj);
-  } else {
-    preProcessorMetrics_.preProcReqInvalid.Get().Inc();
-    delete trueTypeObj;
-  }
+  msgs_.push(trueTypeObj);
+  msgLoopSignal_.notify_one();
 }
 
 template <>
 void PreProcessor::messageHandler<PreProcessReplyMsg>(MessageBase *msg) {
   PreProcessReplyMsg *trueTypeObj = new PreProcessReplyMsg(msg);
   trueTypeObj->setPreProcessorHistograms(&histograms_);
   delete msg;
-  if (validateMessage(trueTypeObj)) {
-    onMessage(trueTypeObj);
-  } else {
-    preProcessorMetrics_.preProcReqInvalid.Get().Inc();
-    delete trueTypeObj;
+  msgs_.push(trueTypeObj);
+  msgLoopSignal_.notify_one();
+}
+
+void PreProcessor::loop() {
+  while (!msgLoopDone_) {
+    int count = 0;
+    {
+      std::unique_lock<std::mutex> l(msgLock_);
+      while (!msgLoopDone_ && msgs_.empty()) msgLoopSignal_.wait(l);
+      if (msgLoopDone_) break;
+      count = msgs_.read_available();
+    }
+
+    while (!msgLoopDone_ && count--) {
+      auto m = msgs_.front();
+      if (validateMessage(m)) {
+        switch (m->type()) {
+          case (MsgCode::ClientBatchRequest): {
+            onMessage<ClientBatchRequestMsg>(static_cast<ClientBatchRequestMsg *>(m));
+            break;
+          }
+          case (MsgCode::ClientPreProcessRequest): {
+            onMessage<ClientPreProcessRequestMsg>(static_cast<ClientPreProcessRequestMsg *>(m));
+            break;
+          }
+          case (MsgCode::PreProcessRequest): {
+            onMessage<PreProcessRequestMsg>(static_cast<PreProcessRequestMsg *>(m));
+            break;
+          }
+          case (MsgCode::PreProcessReply): {
+            onMessage<PreProcessReplyMsg>(static_cast<PreProcessReplyMsg *>(m));
+            break;
+          }
+          default:
+            LOG_ERROR(logger(), "Unknown message" << KVLOG(m->type()));
+        }
+
+      } else {
+        preProcessorMetrics_.preProcReqInvalid.Get().Inc();
+      }
```

Comment:
+ delete msg

Reply:
I think it used internall after that (the same as before, because it is already copy of the original message). Deleting message here leads to crash.